clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -555,6 +555,9 @@ private void close(KafkaChannel channel, boolean processOutstanding) {
 
         channel.disconnect();
 
+        //avoid inconsistent connection states, see KAFKA-5417
+        connected.remove(channel.id());
```

Comment:
IMO, there is not too much difference to do it in "doClose" and "close". Maybe both are ok.
Any channel in "connected" will be reported as connected.
Since it is going to close such channel, it is better to prevent it to be marked as connected than close it after it has been actually marked.